### PR TITLE
[casbin-cpp] Add new port

### LINKED ISSUE #28976: [New Port Request] <casbin-cpp>
Library name:
`casbin-cpp`

Library description:
An authorization library that supports access control models like ACL, RBAC, ABAC in C/C++

Source repository URL:
https://github.com/casbin/casbin-cpp.git

Project homepage:
https://casbin.org/

### PR DESCRIPTION
**Pull request by ex-purple**
- #### What does your PR fix?
  Fixes https://github.com/microsoft/vcpkg/issues/28976

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  all, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

### CODE REVIEW

File: ports/casbin-cpp/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "casbin-cpp",
+  "version-semver": "1.52.0",
```
Comment (LilyWangLL) [inline ports/casbin-cpp/vcpkg.json:3]:
```suggestion
  "version": "1.52.0",
```
Is this port a semver library? If it is not, here should use `"version"`?

File: ports/casbin-cpp/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,23 @@
+vcpkg_from_github(
```
Comment (LilyWangLL) [inline ports/casbin-cpp/portfile.cmake:1]:
```suggestion
# header-only library
vcpkg_from_github(
```

File: ports/casbin-cpp/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    REPO casbin/casbin-cpp
+    REF "v${VERSION}"
+    SHA512 312090ee3c41a138aeefe06544a24d72c6192b1dbb223d83821b790d02c44db23e6f6bc81e693abb0a928820da5feb83bac9c4b6337c15b6bc35325c488d335a
+    HEAD_REF master
+    PATCHES
+        json.patch
+)
+
+vcpkg_cmake_configure(
+    SOURCE_PATH ${SOURCE_PATH}
```
Comment (LilyWangLL) [inline ports/casbin-cpp/portfile.cmake:12]:
```suggestion
    SOURCE_PATH "${SOURCE_PATH}"
```

**Review by LilyWangLL — CHANGES_REQUESTED** [review]

**Review by LilyWangLL — DISMISSED** [review]

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/casbin-cpp/json.patch [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
++++ b/CMakeLists.txt
+@@ -48,7 +48,7 @@ project(
+ # Forbid in-source build.
+ 
+ if(${CMAKE_SOURCE_DIR} STREQUAL ${CMAKE_BINARY_DIR})
+-    message(FATAL_ERROR 
++    message(FATAL_ERROR
+             "In-source build not allowed. Please make a new sub-directory and run CMake from there.")
+ endif()
+
```
[the remark this answers is not in the document]
Comment (ex-purple) [reply ports/casbin-cpp/json.patch]: Removed

File: ports/casbin-cpp/json.patch [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+@@ -79,13 +79,7 @@ endif()
+ # Setting to C++ standard to C++17
+ set(CMAKE_CXX_STANDARD 17)
+ 
+-###############################################################################
+-# Install external dependencies
+-# Some required targets may be created by third-party CMake configs, which 
+-# don't generally produce global targets. To guarantee all imported targets are 
+-# global, this module is included at the project root level.
+-
```
[the remark this answers is not in the document]
Comment (ex-purple) [reply ports/casbin-cpp/json.patch]: Fixed

File: ports/casbin-cpp/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DCASBIN_BUILD_TEST=OFF
+        -DCASBIN_BUILD_BENCHMARK=OFF
+        -DCASBIN_BUILD_PYTHON_BINDINGS=OFF
+)
+
+vcpkg_cmake_install()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
```
[the remark this answers is not in the document]
Comment (ex-purple) [reply ports/casbin-cpp/portfile.cmake]: Fixed

**Review by LilyWangLL — APPROVED** [review]

**Review by BillyONeal — CHANGES_REQUESTED** [review]
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [ ] The generated "usage text" is accurate. See docs/examples/adding-an-explicit-usage.md for context.  
  
This seems wrong: upstream talks about needing to add a library in their readme but currently the usage text is:  
```
casbin-cpp is header-only and can be used from CMake via:

    find_path(CASBIN_CPP_INCLUDE_DIRS "casbin/casbin.h")
    target_include_directories(main PRIVATE ${CASBIN_CPP_INCLUDE_DIRS})
```
Can you investigate? It looks like you might not be actually building the library?
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

**Comment by ex-purple** [comment]
@BillyONeal you are right about a cmake usage. Oddly, casbin-cpp does not install the library and cmake config too. 

Looks like this project orient to use FetchContent. Local installation looks broken. Patches are the only way to fix it.

**Comment by ex-purple** [comment]
@BillyONeal I figured out what's wrong
Library and CMake configuration are installed in the CMake User Package Registry.

**Comment by BillyONeal** [comment]
> @BillyONeal I figured out what's wrong Library and CMake configuration are installed in the CMake User Package Registry.

I don't understand what this means. They are not being installed in anything deployed by vcpkg. If the port is expecting to touch something outside of the installed tree to do its installation, that will fail when the port is restored from the binary cache (where `portfile.cmake` does not run, only the reuslts the port installed to `${CURRENT_PACKAGES_DIR}` are relevant then).

**Comment by ex-purple** [comment]
I understand that this approach is incorrect for vpkg. To fix this, I will need to create a patch.

**Comment by ex-purple** [comment]
Sorry, this port is no more interesting for me.

**Pull request closed without merging**